File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/domain/LegacyToolDraftApplier.kt
package com.rememberber.mootool.next.compose.domain

import com.rememberber.mootool.next.compose.model.ToolId
import com.rememberber.mootool.next.compose.sessions.CodeRunSessionSnapshot
import com.rememberber.mootool.next.compose.sessions.SessionManager
import com.rememberber.mootool.next.compose.sessions.TimeSession
import com.rememberber.mootool.next.compose.storage.HistoryRepository

/**
 * Applies Java-edition `t_func_content` drafts (imported as `operation = draft`) into live tool sessions.
 * History rows still go through [com.rememberber.mootool.next.compose.storage.HistoryRepository].
 */
object LegacyToolDraftApplier {
    fun apply(
        sessionManager: SessionManager,
        rows: List<ImportedLegacyHistory>,
        history: HistoryRepository? = null,
    ): Int {
        val drafts = rows.filter { it.operation == "draft" && it.input.isNotBlank() }
        if (drafts.isEmpty()) return 0

        var applied = 0
        var codeSnapshot = sessionManager.codeRunSession().snapshotState()
        var codeChanged = false
        drafts.forEach { row ->
            val runtimeKey = codeRuntimeDraftKey(row.legacyFunc) ?: return@forEach
            codeSnapshot = patchCodeRunDraft(codeSnapshot, runtimeKey, row.input)
            codeChanged = true
            applied++
        }
        if (codeChanged) {
            sessionManager.codeRunSession().restore(codeSnapshot)
            sessionManager.persistCodeRun()
        }

        applied += applyLast(drafts, "regex") { row ->
            sessionManager.regexSession().source = clip(row.input)
            sessionManager.persistRegex()
        }
        applied += applyLast(drafts, "jsonbeauty", "json") { row ->
            val session = sessionManager.jsonSession()
            val text = clip(row.input)
            session.editor.setText(text, recordUndo = false)
            session.savedText = text
            sessionManager.persistJson()
        }
        applied += applyLast(drafts, "textdiffleft") { row ->
            val session = sessionManager.diffSession()
            session.left = clip(row.input)
            session.result = DiffEngine.compare(session.left, session.right, session.ignoreWhitespace)
            sessionManager.persistDiff()
        }
        applied += applyLast(drafts, "textdiffright") { row ->
            val session = sessionManager.diffSession()
            session.right = clip(row.input)
            session.result = DiffEngine.compare(session.left, session.right, session.ignoreWhitespace)
            sessionManager.persistDiff()
        }
        applied += applyLast(drafts, "timeconvert") { row ->
            applyTimeDraft(sessionManager.timeSession(), row.input, history)
            sessionManager.persistTime()
        }
        applied += applyLast(drafts, "calculator") { row ->
            LegacyCalculatorDraft.apply(sessionManager.calculatorSession(), row.input)
            sessionManager.persistCalculator()
        }
        applied += applyLast(drafts, "qrcode", "qr") { row ->
            val content = LegacyConsoleDraft.extractQrGenerateContent(row.input)
                ?: row.input.lineSequence().firstOrNull { it.isNotBlank() }?.trim()
                ?: row.input.trim()
            sessionManager.qrSession().content = content
            sessionManager.persistQr()
        }

        return applied
    }

    private fun applyLast(drafts: List<ImportedLegacyHistory>, vararg keys: String, block: (ImportedLegacyHistory) -> Unit): Int {
        val wanted = keys.map { it.lowercase() }.toSet()
        val row = drafts.lastOrNull { normalizedFunc(it.legacyFunc) in wanted } ?: return 0
        block(row)
        return 1
    }

    private fun normalizedFunc(legacyFunc: String): String =
        legacyFunc.trim().lowercase().replace("_", "").replace("-", "")

    private fun clip(text: String): String = text.take(CodeRunEngine.MAX_CODE_BYTES)

    private fun codeRuntimeDraftKey(legacyFunc: String): String? {
        val compact = normalizedFunc(legacyFunc)
        return when (compact) {
            "java", "javaconsole", "coderun" -> "java"
            "groovy" -> "groovy"
            "python" -> "python"
            "node", "nodejs" -> "node"
            else -> null
        }
    }

    private fun patchCodeRunDraft(snapshot: CodeRunSessionSnapshot, key: String, content: String): CodeRunSessionSnapshot {
        val text = clip(content)
        return when (key) {
            "java" -> snapshot.copy(javaCode = text)
            "groovy" -> snapshot.copy(groovyCode = text)
            "python" -> snapshot.copy(pythonCode = text)
            "node" -> snapshot.copy(nodeCode = text)
            else -> snapshot
        }
    }

    private fun applyTimeDraft(session: TimeSession, content: String, history: HistoryRepository?) {
        val parsed = LegacyTimeConvertDraft.parse(content)
        if (parsed.isNotEmpty()) {
            parsed.forEach { entry ->
                history?.save(
                    ToolId.TimeConvert.id,
                    entry.summary,
                    entry.summary,
                    entry.input,
                    entry.output,
                    LegacyTimeConvertDraft.optionsJson(entry.zone, entry.unit),
                )
            }
            applyParsedTimeEntry(session, parsed.last())
            return
        }
        val line = content.lineSequence().firstOrNull { it.isNotBlank() }?.trim() ?: return
        if (line.all { it.isDigit() || (line.startsWith("-") && line.drop(1).all { it.isDigit() }) }) {
            session.timestamp = line
            session.unit = if (line.length >= 13) TimestampUnit.Millisecond else TimestampUnit.Second
        } else {
            session.localTime = line
        }
    }

    private fun applyParsedTimeEntry(session: TimeSession, entry: LegacyTimeConvertDraft.ParsedEntry) {
        session.zone = entry.zone.ifBlank { session.zone }
        session.unit = if (entry.unit == "millisecond") TimestampUnit.Millisecond else TimestampUnit.Second
        val input = entry.input.trim()
        val output = entry.output.trim()
        if (input.all { it.isDigit() || (input.startsWith("-") && input.drop(1).all { it.isDigit() }) }) {
            session.timestamp = input
            if (output.isNotEmpty()) session.localTime = output
        } else {
            session.localTime = input
            session.timestamp = output
        }
    }

}

File: next-macos-native/Sources/MooToolNextCore/LegacyToolDraftApplier.swift
import Foundation

public struct ImportedLegacyToolDraft: Equatable, Sendable {
    public var legacyFunc: String
    public var content: String
    public var dedupeKey: String

    public init(legacyFunc: String, content: String, dedupeKey: String) {
        self.legacyFunc = legacyFunc
        self.content = content
        self.dedupeKey = dedupeKey
    }
}

public struct LegacyToolDraftApplyResult: Equatable, Sendable {
    public var drafts: [String: DraftRecord]
    public var history: [HistoryRecord]
    public var applied: Int
}

public enum LegacyToolDraftApplier {
    public static let maxCodeBytes = 512_000

    public static func apply(
        rows: [ImportedLegacyToolDraft],
        merging drafts: [String: DraftRecord],
        existingHistory: [HistoryRecord] = []
    ) -> LegacyToolDraftApplyResult {
        let candidates = rows.filter { !$0.content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        if candidates.isEmpty {
            return LegacyToolDraftApplyResult(drafts: drafts, history: existingHistory, applied: 0)
        }

        var result = drafts
        var history = existingHistory
        var applied = 0

        func patch(_ toolID: String, _ block: (inout DraftRecord) -> Void) {
            var draft = result[toolID] ?? DraftRecord()
            block(&draft)
            result[toolID] = draft
        }

        var codeByMode: [String: String] = [:]
        for row in candidates {
            guard let runtime = codeRuntime(row.legacyFunc) else { continue }
            codeByMode[runtime] = clip(row.content)
        }
        if let lastCode = candidates.last(where: { codeRuntime($0.legacyFunc) != nil }),
           let runtime = codeRuntime(lastCode.legacyFunc) {
            patch("java") {
                $0.mode = runtime
                $0.input = codeByMode[runtime] ?? clip(lastCode.content)
            }
            applied += 1
        }

        if let row = lastRow(candidates, keys: ["regex"]) {
            patch("regex") { $0.input = clip(row.content) }
            applied += 1
        }
        if let row = lastRow(candidates, keys: ["jsonbeauty", "json"]) {
            patch("json") { $0.input = clip(row.content) }
            applied += 1
        }
        if let row = lastRow(candidates, keys: ["textdiffleft"]) {
            patch("textDiff") { $0.input = clip(row.content) }
            applied += 1
        }
        if let row = lastRow(candidates, keys: ["textdiffright"]) {
            patch("textDiff") { $0.secondary = clip(row.content) }
            applied += 1
        }
        if let row = lastRow(candidates, keys: ["timeconvert", "time"]) {
            applied += applyTime(row: row, patch: patch, history: &history)
        }
        if let row = lastRow(candidates, keys: ["calculator", "calc"]) {
            patch("calculator") { LegacyCalculatorDraft.apply(to: &$0, raw: row.content) }
            applied += 1
        }
        if let row = lastRow(candidates, keys: ["qrcode", "qr"]) {
            let content = LegacyConsoleDraft.extractQrGenerateContent(row.content)
                ?? row.content.split(whereSeparator: \.isNewline).map { $0.trimmingCharacters(in: .whitespaces) }.first { !$0.isEmpty }
                ?? row.content.trimmingCharacters(in: .whitespacesAndNewlines)
            patch("qrCode") { $0.input = clip(content) }
            applied += 1
        }

        return LegacyToolDraftApplyResult(drafts: result, history: history, applied: applied)
    }

    private static func applyTime(row: ImportedLegacyToolDraft, patch: (String, (inout DraftRecord) -> Void) -> Void, history: inout [HistoryRecord]) -> Int {
        let parsed = LegacyTimeConvertDraft.parse(row.content)
        if !parsed.isEmpty {
            for entry in parsed {
                var record = HistoryRecord(toolID: "timeConvert", draft: DraftRecord())
                record.draft.input = entry.input
                record.draft.output = entry.output
                record.draft.option = entry.zone
                record.draft.mode = entry.unit
                record.draft.secondary = entry.summary
                history.append(record)
            }
            applyParsedTime(parsed.last!, patch: patch)
            return 1
        }
        patch("timeConvert") { draft in
            let line = row.content.split(whereSeparator: \.isNewline).map { $0.trimmingCharacters(in: .whitespaces) }.first { !$0.isEmpty } ?? ""
            if isIntegerToken(line) {
                draft.input = line
                draft.mode = line.count >= 13 ? "millisecond" : "second"
            } else {
                draft.secondary = line
            }
        }
        return 1
    }

    private static func applyParsedTime(_ entry: LegacyTimeConvertDraft.ParsedEntry, patch: (String, (inout DraftRecord) -> Void) -> Void) {
        patch("timeConvert") { draft in
            if !entry.zone.isEmpty { draft.option = entry.zone }
            draft.mode = entry.unit
            let input = entry.input.trimmingCharacters(in: .whitespacesAndNewlines)
            let output = entry.output.trimmingCharacters(in: .whitespacesAndNewlines)
            if isIntegerToken(input) {
                draft.input = input
                if !output.isEmpty { draft.secondary = output }
            } else {
                draft.secondary = input
                draft.input = output
            }
        }
    }

    private static func lastRow(_ rows: [ImportedLegacyToolDraft], keys: [String]) -> ImportedLegacyToolDraft? {
        let wanted = Set(keys.map(normalizedFunc))
        return rows.last { wanted.contains(normalizedFunc($0.legacyFunc)) }
    }

    private static func normalizedFunc(_ legacyFunc: String) -> String {
        legacyFunc.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            .replacingOccurrences(of: "_", with: "")
            .replacingOccurrences(of: "-", with: "")
    }

    private static func codeRuntime(_ legacyFunc: String) -> String? {
        switch normalizedFunc(legacyFunc) {
        case "java", "javaconsole", "coderun": return "Java"
        case "groovy": return "Groovy"
        case "python": return "Python"
        case "node", "nodejs": return "JavaScript"
        default: return nil
        }
    }

    private static func clip(_ text: String) -> String {
        if text.utf8.count <= maxCodeBytes { return text }
        return String(text.prefix(maxCodeBytes))
    }

    private static func isIntegerToken(_ text: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return false }
        return Int64(trimmed) != nil
    }
}
